test/sharness/t0110-gateway.sh

```diff
@@ -188,7 +188,7 @@ test_expect_success "Add compact blocks" '
   printf "foofoo" > expected
 '
 
-test_expect_success "GET compact blocks succeeds" '
+test_expect_failure "GET compact blocks succeeds" '
```

Comment:
These sorts of changes make me nervous. Whats the technical reason for the swap?

Reply:
That's not good... I assumed that all of the files *we* produced did had blocksizes listed. @whyrusleeping did we change this at some point?

Reply:
@Stebalien I think this test is testing a specifically constructed block.

Reply:
I think we should add a decent number of unit tests for stuff like this. Testing this sort of specific detailed logic via sharness feels wrong.

Reply:
So is the consistence that we should remove this test and instead test that node in a unit test?  And to be clear, do we want this (in my opinion) ill-formed node to fail to validate?

Reply:
This test was added as custom formed blocks might not have sizes but they still should be plain gettable from HTTP. They still can be read sequentially, just efficient seeking is impossible.

Reply:
@Kubuxu so is this still something we want to allow (unseekable blocks)?  We would need some special cases to allow it.

Reply:
Yes

Reply:
@whyrusleeping thoughs?

Reply:
@kevina here is the background on this testcase: https://github.com/ipfs/go-ipfs/issues/4286#issuecomment-336163070

Specifically:

>Additionally - the way my project chunks data, my average blocksize is around 4096 bytes ( with a hard-limit on block sizes at 64k ), so forgoing any optional metadata is critical for me in order to reduce the size of my objects ( as the overhead adds up incredibly fast when one deals with hundreds of millions of objects in a DAG ).

[@whyrusleeping's response](https://github.com/ipfs/go-ipfs/issues/4286#issuecomment-336195370):
> The sizes are optional

@kevina please do not break this relied-upon feature ( neither in PB- nor in the future CBOR-UnixFS )

Reply:
So, this is a problem. Unix programs *expect* files to be seekable and will break horribly if they aren't (and I'm not going to break every single unix program ever written). However, I also don't want to break existing uses if I can help it and this problem doesn't seem insurmountable.

We *should* be able to fall back on dumb seeking (although we may want to disable this from the gateway). Basically, the logic would be as follows:

1. If blocksizes are defined, use them.
2. If a block with children *has no blocksizes* defined (*none* at all, not some, not half, not incorrect ones, **none**), dumb seek through that portion of the file.

This actually gives us the best of both worlds. Files with lots of tiny blocks can put blocksizes at the top layers of the DAG and omit them in the lower levels.

----

In terms of *implementation*, we can punt on dumb seeking for now. Simply validate blocksizes if present and error when seeking when they're not present.

@kevina how hard would this be?

Reply:
I will see what I can do.

Reply:
Done.